hail/python/test/hail/methods/test_impex.py

```diff
@@ -619,6 +619,19 @@ def test_drop(self):
         self.assertEqual(dc._force_count_rows(), 199)
         self.assertEqual(dc._force_count_cols(), 0)
 
+    def test_linear_regression(self):
+        hl.index_bgen(resource('example.8bits.bgen'))
```

Comment:
how does linreg trigger the requiredness bug?

Reply:
can we isolate the bug to something more specific? If we're testing field requiredness, the test should be "test_missing_gp_values" or something, no?

Reply:
It's how I found the problem when it was going through MatrixMapEntries and the code generator on reading the entry field threw an error. I added the test since this is something we want to make sure we can do anyways. `entries().show()` didn't trigger the bug. I can write a different test if you think that would be better.

Reply:
what error did entries().show() throw?

Reply:
er, what error did the original linreg throw without the fix, I mean

Reply:
I'm pushing back because I think we need to be much more diligent about code that makes it into tests - see our struggles with the scala tests as an example of why. 

as written, this is a test that tests over the integration of a bunch of things, but it doesn't test correctness, just that the thing executes successfully. The name doesn't really reflect what it's designed to test, either (the requiredness bug).

We should have integration tests, but IMO that doesn't mean we should let things that can have isolated unit tests become integration tests unnecessarily .

Reply:
I understand your reasoning for not wanting integration tests there. However, I am having a difficult time replicating the bug. I tried some combination of entries.take() on the specific sample, variant pair that I know is missing and they all succeeded. It l

The stack trace was:
```
Java stack trace:
org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 6.0 failed 1 times, most recent failure: Lost task 0.0 in stage 6.0 (TID 6, localhost, executor driver): is.hail.utils.HailException: Required field cannot be missing.
	at is.hail.codegen.generated.C17.apply(Unknown Source)
	at is.hail.codegen.generated.C17.apply(Unknown Source)
	at is.hail.expr.ir.MatrixMapEntries$$anonfun$34$$anonfun$apply$16.apply(MatrixIR.scala:1108)
	at is.hail.expr.ir.MatrixMapEntries$$anonfun$34$$anonfun$apply$16.apply(MatrixIR.scala:1104)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
	at scala.collection.Iterator$$anon$12.next(Iterator.scala:444)
	at is.hail.rvd.OrderedRVD$$anonfun$apply$18$$anon$3.next(OrderedRVD.scala:1106)
	at is.hail.rvd.OrderedRVD$$anonfun$apply$18$$anon$3.next(OrderedRVD.scala:1100)
	at scala.collection.Iterator$$anon$12.next(Iterator.scala:444)
	at is.hail.io.RichContextRDDRegionValue$$anonfun$boundary$extension$1$$anon$1.next(RowStore.scala:1115)
	at is.hail.io.RichContextRDDRegionValue$$anonfun$boundary$extension$1$$anon$1.next(RowStore.scala:1099)
	at scala.collection.Iterator$$anon$12.next(Iterator.scala:444)
	at is.hail.rvd.OrderedRVD$$anonfun$apply$18$$anon$3.next(OrderedRVD.scala:1106)
	at is.hail.rvd.OrderedRVD$$anonfun$apply$18$$anon$3.next(OrderedRVD.scala:1100)
	at scala.collection.Iterator$$anon$12.next(Iterator.scala:444)
	at is.hail.utils.richUtils.RichIterator$$anon$1$$anon$2.next(RichIterator.scala:94)
	at is.hail.methods.LinearRegression$$anonfun$4$$anonfun$apply$2.apply(LinearRegression.scala:87)
	at is.hail.methods.LinearRegression$$anonfun$4$$anonfun$apply$2.apply(LinearRegression.scala:84)
	at scala.collection.Iterator$$anon$12.nextCur(Iterator.scala:434)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:440)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:438)
	at is.hail.rvd.OrderedRVD$$anonfun$apply$18$$anon$3.hasNext(OrderedRVD.scala:1103)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:438)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:438)
	at is.hail.rvd.OrderedRVD$$anonfun$apply$18$$anon$3.hasNext(OrderedRVD.scala:1103)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:438)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:438)
	at is.hail.rvd.OrderedRVD$$anonfun$apply$18$$anon$3.hasNext(OrderedRVD.scala:1103)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:438)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:438)
	at is.hail.rvd.OrderedRVD$$anonfun$apply$18$$anon$3.hasNext(OrderedRVD.scala:1103)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:438)
	at is.hail.utils.package$.getIteratorSizeWithMaxN(package.scala:357)
	at is.hail.sparkextras.ContextRDD$$anonfun$15.apply(ContextRDD.scala:498)
	at is.hail.sparkextras.ContextRDD$$anonfun$15.apply(ContextRDD.scala:498)
	at is.hail.sparkextras.ContextRDD$$anonfun$runJob$1.apply(ContextRDD.scala:528)
	at is.hail.sparkextras.ContextRDD$$anonfun$runJob$1.apply(ContextRDD.scala:526)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2062)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2062)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87)
	at org.apache.spark.scheduler.Task.run(Task.scala:108)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:335)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

Driver stacktrace:
	at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1499)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1487)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1486)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1486)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:814)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:814)
	at scala.Option.foreach(Option.scala:257)
	at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:814)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:1714)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1669)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1658)
	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:48)
	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:630)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2022)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2043)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2062)
	at is.hail.sparkextras.ContextRDD.runJob(ContextRDD.scala:524)
	at is.hail.sparkextras.ContextRDD.head(ContextRDD.scala:498)
	at is.hail.rvd.OrderedRVD.head(OrderedRVD.scala:490)
	at is.hail.rvd.OrderedRVD.head(OrderedRVD.scala:31)
	at is.hail.expr.ir.TableHead.execute(TableIR.scala:331)
	at is.hail.table.Table.value$lzycompute(Table.scala:203)
	at is.hail.table.Table.value(Table.scala:201)
	at is.hail.table.Table.rdd$lzycompute(Table.scala:213)
	at is.hail.table.Table.rdd(Table.scala:213)
	at is.hail.table.Table.collect(Table.scala:568)
	at is.hail.table.Table.showString(Table.scala:622)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:280)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:214)
	at java.lang.Thread.run(Thread.java:745)is.hail.utils.HailException: Required field cannot be missing.
	at is.hail.codegen.generated.C17.apply(Unknown Source)
	at is.hail.codegen.generated.C17.apply(Unknown Source)
	at is.hail.expr.ir.MatrixMapEntries$$anonfun$34$$anonfun$apply$16.apply(MatrixIR.scala:1108)
	at is.hail.expr.ir.MatrixMapEntries$$anonfun$34$$anonfun$apply$16.apply(MatrixIR.scala:1104)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
	at scala.collection.Iterator$$anon$12.next(Iterator.scala:444)
	at is.hail.rvd.OrderedRVD$$anonfun$apply$18$$anon$3.next(OrderedRVD.scala:1106)
	at is.hail.rvd.OrderedRVD$$anonfun$apply$18$$anon$3.next(OrderedRVD.scala:1100)
	at scala.collection.Iterator$$anon$12.next(Iterator.scala:444)
	at is.hail.io.RichContextRDDRegionValue$$anonfun$boundary$extension$1$$anon$1.next(RowStore.scala:1115)
	at is.hail.io.RichContextRDDRegionValue$$anonfun$boundary$extension$1$$anon$1.next(RowStore.scala:1099)
	at scala.collection.Iterator$$anon$12.next(Iterator.scala:444)
	at is.hail.rvd.OrderedRVD$$anonfun$apply$18$$anon$3.next(OrderedRVD.scala:1106)
	at is.hail.rvd.OrderedRVD$$anonfun$apply$18$$anon$3.next(OrderedRVD.scala:1100)
	at scala.collection.Iterator$$anon$12.next(Iterator.scala:444)
	at is.hail.utils.richUtils.RichIterator$$anon$1$$anon$2.next(RichIterator.scala:94)
	at is.hail.methods.LinearRegression$$anonfun$4$$anonfun$apply$2.apply(LinearRegression.scala:87)
	at is.hail.methods.LinearRegression$$anonfun$4$$anonfun$apply$2.apply(LinearRegression.scala:84)
	at scala.collection.Iterator$$anon$12.nextCur(Iterator.scala:434)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:440)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:438)
	at is.hail.rvd.OrderedRVD$$anonfun$apply$18$$anon$3.hasNext(OrderedRVD.scala:1103)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:438)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:438)
	at is.hail.rvd.OrderedRVD$$anonfun$apply$18$$anon$3.hasNext(OrderedRVD.scala:1103)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:438)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:438)
	at is.hail.rvd.OrderedRVD$$anonfun$apply$18$$anon$3.hasNext(OrderedRVD.scala:1103)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:438)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:438)
	at is.hail.rvd.OrderedRVD$$anonfun$apply$18$$anon$3.hasNext(OrderedRVD.scala:1103)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:438)
	at is.hail.utils.package$.getIteratorSizeWithMaxN(package.scala:357)
	at is.hail.sparkextras.ContextRDD$$anonfun$15.apply(ContextRDD.scala:498)
	at is.hail.sparkextras.ContextRDD$$anonfun$15.apply(ContextRDD.scala:498)
	at is.hail.sparkextras.ContextRDD$$anonfun$runJob$1.apply(ContextRDD.scala:528)
	at is.hail.sparkextras.ContextRDD$$anonfun$runJob$1.apply(ContextRDD.scala:526)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2062)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2062)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87)
	at org.apache.spark.scheduler.Task.run(Task.scala:108)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:335)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

Reply:
will `mt._typecheck()` catch it?

Reply:
will `mt._typecheck()` catch it?

Reply:
What is the IR there? That error looks like the one that gets thrown trying to return a missing value from a compiled function, rather than having to do with requiredness.

Reply:
this has reminded me how much I hate that LinearRegression produces a MT, not a table. It's going to really kill performance.

Reply:
So I did this test the select entries triggers the type check to fail:

```
    var mt = hc.importBgens(Array(bgenFile), sampleFile = Some("src/test/resources/example.sample"))
    mt.typecheck() # Passes!
    mt = mt.selectCols("(InsertFields (Ref sa) (y (ApplySeeded rand_unif 7370157127079664914 (Apply toFloat64 (I32 0)) (Apply toFloat64 (I32 1)))) (pc1 (ApplySeeded rand_unif 6181994927155295486 (Apply toFloat64 (I32 0)) (Apply toFloat64 (I32 1)))))", None)
    mt.typecheck() # Passes!
    mt = mt.selectEntries("(InsertFields (Ref g) (__uid_1 (Apply toFloat64 (Apply nNonRefAlleles (GetField GT (Ref g))))))")
    mt.typecheck() # Fails!
```

Reply:
so further conversation with @tpoterba---there's an argument to be made here that since everything is computed from GP, the requiredness is correct here because if the GP is missing in the BGEN file, then we consider the entire entry to be missing (but if the GP exists, then the dosage must exist, so the entry can either be entirely missing or contain a dosage)

The other fix for this is to change the semantics of InsertFields to preserve the missingness of the original struct. This is nice for stuff like filtering entries, where an `annotate_entries` would otherwise unfilter previously filtered entries. This means that the missing entry in this case would never try to explode out the (required) dosage, since the resulting entry would still be missing.

Reply:
after talking with Amanda I do think I'm convinced that the requiredness of the BGEN types isn't wrong.

I think we should make a test that isolates this behavior to InsertFields (no BGEN or regression required) and fix that.